Hero Header P1

danielfmiranda opened the pull request:
Closes #8370

First pass at P1 of the interactive report template hero, in this PR we have:

- Made Authors Optional
- Gave users the choice of old article layout + 3 new ones (half screen left / half screen right / full screen)
- Gave users option to upload image or video for hero section media
- Gave users option to upload custom button + choose button style from (primary/secondary/tertiary)


**Some notes:**

- The play/pause button on hero videos is not working yet due to the article sticky nav, this will be taken care of with the new sticky nav.
- Since we are adding custom colors in p2, I have not implemented the ability to choose between black and white text. That is because if we added the "black/white" field now, it would no longer be needed shortly after.
- This PR does not include the help panels, just text, as I am anticipating some design feedback and would like to have the help panels include updated photos if we still want to include them. 

**Steps to test:**

1. Please visit any of the test pages below to take a look at the new formats. 
2. If you would like, please visit the review app cms using the credentials "review/review" and create a test article page of your own. 
3. Please leave any feedback in this ticket
4. Testing is complete! 


**Link to sample test page**:

- [Updated Hero - Full Screen Image - Primary Button - Custom Icon](https://foundation-s-hero-heade-iruhbp.mofostaging.net/en/publication-page-with-child-article-pages/fixed-title-article-page/)
- [Updated Hero - Full Screen Video - Primary Button - Custom Icon ](https://foundation-s-hero-heade-iruhbp.mofostaging.net/en/publication-page-with-child-article-pages/financial-hotel-inside-include-positive-water-budget/)
- [Updated Hero - Image Left - Secondary Button - Default Icon ](https://foundation-s-hero-heade-iruhbp.mofostaging.net/en/publication-page-with-child-article-pages/range-newspaper-south-own-between-music-knowledge/)
- [Updated Hero - Video Left - Tertiary Button - Custom Icon](https://foundation-s-hero-heade-iruhbp.mofostaging.net/en/publication-page-with-child-article-pages/war-still-fill-when-firm/)
- [Updated Hero - Image Right - Secondary Button - Custom Icon](https://foundation-s-hero-heade-iruhbp.mofostaging.net/en/publication-page-with-child-article-pages/hot-break-by-step/)
- [Updated Hero - Video Right - Tertiary Button - Default Icon](https://foundation-s-hero-heade-iruhbp.mofostaging.net/en/publication-page-with-child-article-pages/coach-bar-chair-service-detail/)
- [Original Hero](https://foundation-s-hero-heade-iruhbp.mofostaging.net/en/publication-page-with-child-article-pages/certainly-bit-single-fly-recognize-trade/)

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/article_page.html:
```diff
@@ -11,14 +11,18 @@
 
 {% block content %}
 
+{% if page.hero_layout != "static" %}
```
If desired page hero layout is not "static" aka the old template, render the new customizable template

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/article_hero_guts.html:
```diff
@@ -0,0 +1,90 @@
+   {% endwith %}
+  {% endif %}
+
+</div>
```
```suggestion
</div>

```

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/article_hero.html:
```diff
@@ -0,0 +1,73 @@
+  </div>
+
+{% endif %}
+{% endwith %}
```
```suggestion
{% endwith %}

```

danielfmiranda commented on source/js/foundation/template-js-handler/article-hero-video-handler.js:
```diff
@@ -0,0 +1,65 @@
+   */
+  export default () => {
+    articleHeroVideoHandler();
+  };
```
```suggestion
  };

```

danielfmiranda reviewed (commented):
Updating newlines at end of files

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/article_hero.html:
```diff
@@ -0,0 +1,73 @@
+{% with control_btn_classes="ml-auto tw-opacity-80 hover:tw-opacity-100 tw-hidden" full_screen_content_classes="tw-absolute tw-min-w-full tw-min-h-full tw-max-w-none tw-top-0" two_col_content_classes="tw-h-full tw-w-full tw-object-center tw-object-cover" %}
+
+
+{% if page.hero_layout != "full_screen" %}
```
This template handles all 3 of the updated article hero layouts (full screen/Image left/image right)

If the layout choice is either "image left" or "image right" or more simply "not full screen," then render a two column responsive layout.


The "else" block below renders the 1 column, full screen image/video layout.

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/article_hero_guts.html:
```diff
@@ -0,0 +1,90 @@
+{% load static i18n wagtailimages_tags homepage_tags%}
+{% get_page_authors page as authors %}
+
```
This file follows the other article/publication file pretty closely with some additional if statements for whether or not the user would like to render some things such as the authors. However in preparation for the P2 asks which are custom colors and such, I think this is best in its own file so we can update it accordingly in the future without making any changes to the original article/publication header file

danielfmiranda commented on source/js/foundation/template-js-handler/article-hero-video-handler.js:
```diff
@@ -0,0 +1,66 @@
+ * Adding play/pause functionality to the optional Article Page hero video section.
+ */
+
+ const articleHeroVideoHandler = () => {
```
This file looks for a video + the play and pause buttons. If it finds them, render the pause button and play the video automatically. 

We also add a click handler on the entire video to pause it.

*Note*: At the moment, a bug with the original article pages "sticky nav" is preventing the pause/play button to be clicked. However after some discussion with Simon, we think it would be best to work together to fix this bug since he is implementing a new, updated sticky nav.

Pomax replied:
If we don't need to preserve the declare-time `this`, we don't need arrow functions.

```suggestion
function articleHeroVideoHandler() {
```

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/publications/article.py:
```diff
@@ -88,11 +102,47 @@ class ArticlePage(FoundationMetadataPageMixin, Page):
+        on_delete=models.SET_NULL,
+        related_name='+',
+        help_text="Custom Icon for download button, please use http://feathericons.com"
+
```
rogue newline~

Pomax commented on network-api/networkapi/wagtailpages/pagemodels/publications/article.py:
```diff
@@ -88,11 +102,47 @@ class ArticlePage(FoundationMetadataPageMixin, Page):
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name='+',
+        help_text="Custom Icon for download button, please use http://feathericons.com"
```
`https`, at the very least, but I'm a little confused why a download icon is a foreignkey to a Document, what are we actually doing here?

danielfmiranda replied:
Hi @Pomax, 

The request from design was to have users be able to update the page's download button with a custom icon from feathericons.com.

FeatherIcons are only downloadable in SVG which appear to be incompatible with the imageField, which I believe would have been the ideal use case.

After some further investigation, I found [this comment ](https://www.reddit.com/r/WagtailCMS/comments/eey2wn/comment/fh6jtvu/?utm_source=share&utm_medium=web2x&context=3)from Kalob where he explains the limitations of SVG and the wagtail ImageField, and a workaround in order to upload them to the CMS. 

> SVG's aren't technically images, they're xml-like files what get read and rendered by a browser (or other design program like Photoshop or Illustrator).
> Because they are files, they need to be uploaded as a file, not an image.
> In Wagtail this goes through either a `models.FileField()` Page property, or through the document uploader and manager.
```
from wagtail.core.models import Page
from wagtail.documents.edit_handlers import DocumentChooserPanel 

class YourPage(Page):
    svg = models.ForeignKey(
        'wagtaildocs.Document',
        blank=True, # or False 
        null=True, # or False 
        related_name='+',
        on_delete=modules.SET_NULL, # Only works with null=True
    )

    content_panels = Page.content_panels + [
        DocumentChooserPanel("svg"),
    ]
```



From Kalob's explanation, it appears that there 2 ways to be able to achieve this:

1. The first one being to use the document manager as we are here, which in my thinking was the most similar to using the image manager.
2. The second option Kalob mentioned was using a filefield, which come to think of it, might be a better option as that would provide a direct file selector instead of using the document manager. For example, here is what the FileField looks like in the "News" Model's thumbnail field:
<img width="1001" alt="Screen Shot 2022-03-22 at 5 41 35 PM" src="https://user-images.githubusercontent.com/18314510/159599812-5aed8da3-a334-49e5-b743-56418316f264.png">

A possible third option, would be to use the [wagtailSVG](https://pypi.org/project/wagtailsvg/) package, the last update to it appears to be Feb 2022, however it also seems to be a small project with 26 stars.  

Let me know if that helped answer the question, or if you would like me to implement this a different way 😄 thanks!

Pomax replied:
Ah, that makes sense. It would probably be good to explain why a Document is used in a code comment here.

Pomax commented on source/js/foundation/template-js-handler/article-hero-video-handler.js:
```diff
@@ -0,0 +1,66 @@
+  /**
+   * Bind handlers to article hero video
+   */
+  export default () => {
+    articleHeroVideoHandler();
+  };
```
There is no reason to wrap this function in another function =)

```suggestion
  export default articleHeroVideoHandler
```

Although as the only thing in this file, it's even better to just start it as 

```js
export default function articleHeroVideoHandler() {
  ...
}
```

Pomax commented on source/js/foundation/template-js-handler/article-hero-video-handler.js:
```diff
@@ -0,0 +1,66 @@
+ */
+
+ const articleHeroVideoHandler = () => {
+    let articlePageHero = document.querySelector(".article-hero");
```
just as a good habit: if the variable is not intended to be reassigned, you'll want to use `const` rather than `let`.

danielfmiranda replied:
Thanks for the catch Pomax, I will remember to do that moving forward! Updated and marking as resolved.

Pomax commented on source/js/foundation/template-js-handler/article-hero-video-handler.js:
```diff
@@ -0,0 +1,66 @@
+    function toggleVideo() {
+        video.paused ? video.play() : video.pause();     
+    }
+
+
+
```
this is a lot of whitespace, suggesting that `prettier` isn't picking it up. Let's make sure it does.

beccaklam reviewed (changes requested):
Hi @danielfmiranda! The header is looking great as I've already told you multiple times 💯 🙌 
I've listed out some change requests below. I'm happy to jump on a call/chat on slack anytime so just let me know if you need more clarification about any of these things!

## CMS:
- [ ] Hero Video: Can we add instructions on how to upload a video?

## Article Page:
- [ ] If there are no H2 anchors and therefore no Article Summary box can we have the content start below the hero header. Perhaps with 48px of space in between the header and the beginning of the content? 
![Screen Shot 2022-03-23 at 2 08 40 PM](https://user-images.githubusercontent.com/20977376/159796082-d074431d-be19-4896-84e5-5d1df2e03d6f.png)

## Mobile
- [x] Sorry this is my fault! The H1 for the header should just be the plain Mobile Heading 1 and not have the over ride of being bold. This is my fault for not cleaning this up in the design file. 
![Screen Shot 2022-03-23 at 3 02 45 PM](https://user-images.githubusercontent.com/20977376/159803474-cbd93546-9125-4c71-8f63-73ade5a509f4.png)

## Hero Header
### For large (desktop) view port sizes:
- [x] Would we be able to increase the padding for the div holding the hero header content? For desktop, or from inspect it looks like the class is _.large\:tw-px-7_  can we increase the padding from 3rem to 3.75rem? 
![Screen Shot 2022-03-23 at 3 11 13 PM](https://user-images.githubusercontent.com/20977376/159804637-3740d846-38a0-4626-aec2-28b06c88e716.png)

- [x] Could we also add a max width of 650px for the hero header content itself (I think the class is _.article-hero-content_)?  
![Screen Shot 2022-03-23 at 3 12 12 PM](https://user-images.githubusercontent.com/20977376/159804671-3274d654-b552-4abd-b83a-7464f923543e.png)

Happy to chat this through on a call if that's easier! Also, if there are better ways to code what I'm describing above please do that instead! Medium and small view ports look fine so those don't need any changes I think 👍

### On the fence:
- [ ] Removing Image Right layout: Kristina has mentioned perhaps removing this layout because the image can get really pushed down on mobile. Up until this point I've wanted to keep the layout but seeing it on staging really does show how the image can disappear below the fold on mobile if there's a lot of header content. That, plus how great Image Left is looking, is making me reconsider my stance. We want to make the template as easy to use as possible for staff so I am leaning towards taking it out but wanted to see if you had any thoughts first?

Pomax reviewed (approved):
some design issues to work through, but from a code perspective this looks good to me! (and also, it just _looks good_ =D)

danielfmiranda commented:
Hi @beccaklam! Thank you very much for the great feedback 😄

To respond to your requests:

## CMS:
I will add some instructions for the video uploading! 

## Article Page:
My Mistake! I will be sure to add some padding for when the summary box is not rendered

## Mobile
No problem! This can be easily removed and the heading can stay one font size.

## Hero Header
### For large (desktop) view port sizes:
No problem! I will update the padding from 3em to the requested size, also, I will implement the max width of 650px.

### Removing Image Right layout:
I am neutral on this, as I do think the summary box looks good when overlapping the video/image. Though I know that the original design meant to have the summary box showing on the initial loading of the page. 

That being said, the question of can we remove it is a yes! It would be as simple as me deleting a line from the code. 

So if the design team decides its best to leave it out, please let me know and I can remove it no problem!

danielfmiranda commented:
Hi @beccaklam, I have implemented the feedback and re-requested review. Thank you!

kristinashu commented:
Wow these are looking so amazing 🙌 🙌 🙌

### Video pause/play

The videos are lovely and the pause works if I click somewhere on the video but it don't work if I click the actual pause/play button.
- [ ] Make the pause/play button clickable

### Download link

@beccaklam did you design a hover state for the download link? Right now on hover it turns bold which feels a little off. Maybe it should change color?

<img width="176" alt="Screen Shot 2022-03-25 at 9 03 32 AM" src="https://user-images.githubusercontent.com/1068783/160158632-985ac911-fa6b-42e2-8a5a-db4b89669688.png">

beccaklam reviewed (changes requested):
Hi @danielfmiranda, thank you for making those changes! These are the remaining changes we talked about, after this we should be good to merge.  Thank you!

## CMS:
### Download Button
- [x] Move 'Download button style' and 'Download button icon' to be grouped with 'Article file'
- [x] Can we change 'Article file' to 'Download button file' to be consistent in references to the button?

## Full Screen Layout:
### Mobile
- [x] Update hard coded padding top and bottom values for the Mobile Full Screen Hero Header layout

danielfmiranda commented:
Hi @kristinashu @beccaklam thank you both for the great feedback!

I have made the following changes: 
## Play/Pause Button Bug:
I have removed the play/pause button for now, as at the moment a bug with the original sticky nav is preventing it from being clickable in the desired location. Since Simon is working on the updated sticky nav, we are going to work together to implement the play pause button alongside it. However, the click to play/pause anywhere on the video remains! 

## CMS:
- I have grouped together the desired fields
- The label for the download file has been changed from "Article File" to "Download Button File"

## Full Screen Layout - Mobile:
- I have updated the padding to be closer to the top for fullscreen pages on mobile.

Please let me know if there are any questions or other updates, thank you!

beccaklam commented:
Hi @danielfmiranda! The changes you made look good! No worries about the play button, Simon had already informed me about that bug. 

Really sorry, but just one more thing, as Kristina mentioned, would we be able to update the Tertiary button hover state? It's my fault for missing it during the design phase but I've done a quick design now. Basically, can we reduce the opacity of both the button and icon to 50% when the text is black, and to 65% when the text is white (though, white text might be a P2?). I've included a design below and it's also in the [Figma file](https://www.figma.com/file/P06Q7hWuILgc9g19AXcP39/Publications-template?node-id=56%3A10279):
<img width="614" alt="Screen Shot 2022-03-25 at 1 43 03 PM" src="https://user-images.githubusercontent.com/20977376/160199038-1711911e-5696-4f0c-a493-11651bfef9b1.png">

beccaklam reviewed (approved):
Thank you for making all those changes @danielfmiranda! This is good to go from design! 🚢 🥳

fessehaye reviewed (approved):
Looks great!
